rp2: Fix linker issue with gap alignment on CI.

dpgeorge opened the pull request:
### Summary

Rp2 CI recently broke, I think due to the runner being updated from `ubuntu-24.04.01` to  `ubuntu-24.04.02`.  GCC remained unchanged but maybe the linker changed.

The CI fails like this:
```
[ 97%] Building C object CMakeFiles/firmware.dir/home/runner/work/micropython/micropython/micropython_repo/lib/pico-sdk/src/rp2_common/cmsis/stub/CMSIS/Device/RP2040/Source/system_RP2040.c.o
[ 98%] Linking CXX executable firmware.elf
ERROR: Cannot plug gap greater than alignment - gap 253876, alignment 32
make[3]: *** [CMakeFiles/firmware.dir/build.make:7195: firmware.elf] Error 253
make[3]: *** Deleting file 'firmware.elf'
make[3]: Leaving directory '/home/runner/work/micropython/micropython/micropython repo/ports/rp2/build-RPI_PICO'
make[2]: *** [CMakeFiles/Makefile2:2228: CMakeFiles/firmware.dir/all] Error 2
make[2]: Leaving directory '/home/runner/work/micropython/micropython/micropython repo/ports/rp2/build-RPI_PICO'
make[1]: *** [Makefile:91: all] Error 2
make[1]: Leaving directory '/home/runner/work/micropython/micropython/micropython repo/ports/rp2/build-RPI_PICO'
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [Makefile:65: all] Error 1
make: Leaving directory '/home/runner/work/micropython/micropython/micropython repo/ports/rp2'
```

### Testing

CI will test this fix.

dpgeorge commented:
I really don't know what is wrong here nor how to fix it.

malcolm-sparkfun commented:
A possibly tangentially related issue: https://github.com/raspberrypi/pico-sdk/issues/2391#issuecomment-2772486582. Seems like GitHub updated the ubuntu runner to install cmake==4.0.0 and it is leading to many CI failures: https://github.com/actions/runner-images/issues/11926.

lurch commented:
Looks like this might be related to https://github.com/raspberrypi/picotool/pull/213 ? ping @will-v-pi

dpgeorge commented:
> Looks like this might be related to [raspberrypi/picotool#213](https://github.com/raspberrypi/picotool/pull/213) ?

Oh, indeed it does!  The error message fits exactly.

It would be nice if picotool printed out something like `picotool: error: ...` so that it's easier to tell where the error comes from.

I guess this needs to be fixed in picotool.

But this brings up another question: why does pico-sdk default to using the `develop` branch of picotool?  That doesn't seem right.  Here's the relevant extract from `tools/Findpicotool.cmake`:
```cmake
        FetchContent_Populate(picotool QUIET
            GIT_REPOSITORY https://github.com/raspberrypi/picotool.git
            GIT_TAG develop

            SOURCE_DIR ${picotool_INSTALL_DIR}/picotool-src
            BINARY_DIR ${picotool_INSTALL_DIR}/picotool-build
            SUBBUILD_DIR ${picotool_INSTALL_DIR}/picotool-subbuild
        )
```
Maybe the `GIT_TAG develop` should be `GIT_TAG master`.

dpgeorge committed:
tools/ci.sh: Manually install picotool for rp2 builds.

If picotool is not installed, it's fetched and built when compiling each
rp2 board.  And the "develop" branch of picotool is used instead of a
release.  Installing it manually using the "master" branch means the latest
released version is used (instead of a possibly unstable development
version), and also makes building each rp2 board a little faster.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
The CI here is now working again, with the fix being to manually install a release version of `picotool`, rather than using the automatically-fetched `develop` branch.

I think we can move ahead with this solution for now in the CI.

For users building the rp2 port, they have a choice to either:
1. Install `picotool` manually (see the README of that project).
2. Edit `lib/pico-sdk/tools/Findpicotool.cmake` and change `GIT_TAG develop` to `GIT_TAG master`.

will-v-pi commented:
> ```
> ERROR: Cannot plug gap greater than alignment - gap 253876, alignment 32
> ```
> 
> Looks like this might be related to [raspberrypi/picotool#213](https://github.com/raspberrypi/picotool/pull/213) ? ping @will-v-pi

Yes, this change was to plug gaps between sections within a segment in the ELF file, as those are incompatible with signing/encrypting which works on entire segments.

Could you send over the ELF files from the failing builds, or the output from `readelf -e` on them, to help me understand what's causing this issue so I can fix it in `picotool`?

dpgeorge commented:
> Found the bug and fixed it

Very good!

dpgeorge commented:
> I'd missed handling for sections which weren't in any segments, so it was returning true for `heap` and `stack_dummy` being in the same segment, which was causing the error.

I can confirm that the latest `develop` branch of `picotool` now allows MicroPython to build (but I didn't test the firmware).

lurch commented:
> 2\. Edit `lib/pico-sdk/tools/Findpicotool.cmake` and change `GIT_TAG develop` to `GIT_TAG master`.

Just FYI, there's now a PR for this https://github.com/raspberrypi/pico-sdk/pull/2401
